### PR TITLE
Attempt one at namespacing the application

### PR DESCRIPTION
**Pull request by wilsonge**

### CODE REVIEW

File: libraries/src/application/Administrator.php [hunk 1]
```diff
@@ -7,16 +7,30 @@
  * @license     GNU General Public License version 2 or later; see LICENSE.txt
  */
 
+namespace Joomla\Cms\Application;
+
 defined('JPATH_PLATFORM') or die;
 
 use Joomla\Registry\Registry;
+use \JInput;
```
Comment (mbabker) [inline libraries/src/application/Administrator.php]: That's just lazy :stuck_out_tongue:

File: libraries/vendor/composer/autoload_classmap.php [hunk 2]
```diff
@@ -6,10 +6,772 @@
 $baseDir = dirname(dirname($vendorDir));
 
 return array(
+    'Cache_Lite' => $vendorDir . '/pear/cache_lite/Cache/Lite.php',
```
Comment (mbabker) [inline libraries/vendor/composer/autoload_classmap.php]: We committing dev resources to repo here?  If not, run `composer install --no-dev` to dump the dev deps.

File: libraries/src/application/Helper.php [hunk 3]
```diff
@@ -7,14 +7,22 @@
...
+use \JInstaller;
+use \JText;
+
 /**
  * Application helper functions
  *
  * @since  1.5
  */
-class JApplicationHelper
+class Helper
```
Comment (mbabker) [inline libraries/src/application/Helper.php]: Even though it's namespaced, I'd suggest not using vague class names like `Helper`.  Gets awkward when you start loading the application helper, component helper, and installer helper in one class, for example.

File: libraries/src/application/Administrator.php (hunk 1 above)
Comment (wilsonge) [inline libraries/src/application/Administrator.php]: It's fine for now. I want to do things one class at a time.

File: libraries/vendor/composer/autoload_classmap.php (hunk 2 above)
Comment (wilsonge) [inline libraries/vendor/composer/autoload_classmap.php]: No but I needed phpunit because I haven't got travis for this repo :P

**Comment by mbabker** [comment]
You should have uppercase folder names once you're in the `libraries/src` folder, so `libraries/src/Application`.

File: libraries/src/application/Helper.php (hunk 3 above)
Comment (nikosdion) [inline libraries/src/application/Helper.php]: I disagree. Your namespace is Application, you should not add the same identifier in the class immediately below it. This is why the `use` construct has the `as` keyword. If we don't follow that rule we will end up with equally pointless class names like `Acme\Example\Admin\Model\AcmeExampleAdminModelFoo`.

**Comment by wilsonge** [comment]
Yah. Good spot. Fixed

**Comment by mbabker** [comment]
I don't know how you're managing the Composer stuff on your local but you've gone and screwed that up six ways from Sunday.

Get your composer.lock back to staging.  Then run `composer install` and the autoloader regenerates, DO NOT RUN `composer update`.  You've effectively updated the entire composer.lock file and all the external dependencies just to add a new namespace to the path lookup.

**Comment by wilsonge** [comment]
Kinda done. Session and Event are still dead tho :/

**Comment by wilsonge** [comment]
Turns out i had the packages i just forgot to call `git add` :P ahh well - ty

File: libraries/vendor/composer/autoload_classmap.php (hunk 2 above)
Comment (nibra) [inline libraries/vendor/composer/autoload_classmap.php]: Am 12.07.2015 um 18:54 schrieb Michael Babker:

> We committing dev resources to repo here?  If not, run `composer install --no-dev` to dump the dev deps.

Add /libraries/vendor to your global .gitignore.

Comment (mbabker) [inline libraries/vendor/composer/autoload_classmap.php]: We still have to get the CMS there.  When we merged Composer resources at 3.4 we decided to commit production code in the vendor folder.  There were two arguments for this; enabling contributors to clone the repo without needing to use extra tools too (Composer) and packaging.  Our patch packages are built by comparing git tags and if the folder isn't tracked in VCS then an alternate approach is needed to decide what to include from vendor.  FWIW in Mautic we use a similar packaging technique and basically take an all-or-nothing approach to it (if .lock is changed include the whole folder).

**Comment by nibra** [comment]
Wouldn't it be more efficient (and better readable) to say
    use Joomla\Cms\Application\Helper as JApplicationHelper;
and leave the rest alone?

**Comment by nikosdion** [comment]
That would work. The only possible downside is missing a replacement somewhere (forget to add the use statement). If you are concentrated and methodical this risk is minimized.

**Comment by wilsonge** [comment]
OK this is complete. Tests pass - just don't try and run the different groups in isolation :P. There is some coupling between the JApplicationCms tests and the tests in JHtml that I couldn't figure out.

**Comment by wilsonge** [comment]
OK there's a bunch of unit tests that seem to be coupled to what's happening in `JApplicationCms`. I can't figure out what's going on. I'm not sure atm if it's related to `JUri` or `JApplicationCms::getTemplate`. I can't see any couplings at the moment. I tried redeclaring the server variables that we were using in `JApplicationCms` tests in case `JUri` was falling over them - but no joys.

I'm really all out of ideas at this point.....

File: libraries/src/application/Helper.php (hunk 3 above)
Comment (mbabker) [inline libraries/src/application/Helper.php]: Generally I agree with that thinking, I just personally dislike the thought of namespaced class names that end up as `Helper` or `Site`.  Even with `as` references, it just doesn't feel right to me.

**Comment by mbabker** [comment]
I took a minimalistic approach in #4 and didn't have the colossal test explosions in JHtml.  That entire package's tests need a lot of attention in general.  I have a gut feeling something you did with the namespacing in the test suite messed with something else.

File: libraries/src/application/Helper.php (hunk 3 above)
Comment (nikosdion) [inline libraries/src/application/Helper.php]: I think consistency is the primary goal of namespaces. You should have things ending in Helper or Site. Take a look at things like Guzzle and Amazon AWS SDK for PHP. They're not at all shy on following this approach.

**Comment by wilsonge** [comment]
OK Well let's go with #4 for now then :) I guess we're going to require refactoring all our tests much earlier on than I hoped

**Pull request closed without merging**